Cache Playground assets to enable offline support

bgrgicak opened the pull request:
## Motivation for the change, related issues

The promise of Playground is WordPress anywhere. 
Playground can run in the browser without a server, but it still requires the internet on every load.

This PR adds caching of Playground assets and ensures they are available when offline.

## Implementation details

The Remote Service Worker will start caching when initialized and clean previous versions if they exist.

The WorkerCache class uses browser cache to preload data when the service worker is loaded.
On each unscoped fetch request, it will attempt to load data from the cache by using the `cachedFetch` method.

Only URLs that aren't PHP scripts (end with _.php_) and valid domains will be cached. (see `isValidHostname` for details)

## Testing Instructions (or ideally a Blueprint)

- Follow the [Test offline support instructions](https://github.com/WordPress/wordpress-playground/pull/1535/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R134)
- If you previously loaded the site unregister the service worker and clear cache storage
- After loading Playground **only once** in your browser turn off the server
- In Dev tools > Application > Service Worker set it to Offline (this ensures remote assets are loaded)
- Refresh the page
- Confirm that Playground loaded
- Browse the site, go to WP-admin, make changes...
- It should all work as usual

**Note: In Chrome the offline checkbox sometimes stops working. When this happens Playground will load without remote assets and admin styles (including the admin bar) will look broken. If it happens you can confirm it by running `window.navigator.onLine`. To resolve it uncheck and check the Offline checkbox and reload.**

bgrgicak commented:
Playground can now load from the cache when there is no server. 

There is still an issue on the first load where some static assets aren't pulled from cache.
Also, assets are cached only on the second load.

Other TODOs:
- Store cache with a version prefix
- Flush cache when version changes

adamziel commented on packages/php-wasm/web-service-worker/src/initialize-service-worker.ts:
```diff
@@ -13,6 +15,14 @@ import { getURLScope, isURLScoped, setURLScope } from '@php-wasm/scopes';
 
+	self.addEventListener('install', (event) => {
+		try {
+			event.waitUntil(preCacheResources());
```
I saw line targets `index.html`, `sw.js`, and `/` – why those files specifically and not also, say, `style.css`? Also, using `waitUntil` means the service worker won't start serving other requests until those files are cached – was that intended?

bgrgicak replied:
If we don't use `event.waitUntil` we can't ensure all the files will be available for offline mode after the initial load, but I don't think this is a big problem and in my testing, it works without `event.waitUntil`.

bgrgicak replied:
> I saw line targets index.html, sw.js, and / – why those files specifically and not also, say, style.css?

This is the root cause of _There is still an issue on the first load where some static assets aren't pulled from cache._. All files should be preached for offline support to work after the first load. 

I'm searching for a solution now.

bgrgicak replied:
This is working now. I had to preload all assets during `install` to ensure they are available while offline.

~~I'm not happy with the implementation, it requires two fetches to get the asset list, and prel all assets.~~

On one had this is more robust and allows us to do things like change PHP versions while offline, but I would still like to reduce the list size.

adamziel commented on packages/php-wasm/web-service-worker/src/initialize-service-worker.ts:
```diff
@@ -22,28 +32,28 @@ export function initializeServiceWorker(config: ServiceWorkerConfiguration) {
+			!isURLScoped(url) &&
+			!(
+				event.request.referrer &&
+				isURLScoped(new URL(event.request.referrer))
```
Referrer may not be a valid URL and we still need that try/catch:

> The referrer read-only property of the [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request) interface is set by the user agent to be the referrer of the Request. (e.g., `client`, `no-referrer`, or a URL.)

https://developer.mozilla.org/en-US/docs/Web/API/Request/referrer

bgrgicak replied:
Thanks! I didn't know that [c34fd37](https://github.com/WordPress/wordpress-playground/pull/1535/commits/c34fd37ca20b4990899915df479c8e6357fdb815)

adamziel commented on packages/php-wasm/web-service-worker/src/worker-caching.ts:
```diff
@@ -0,0 +1,86 @@
+};
+
+export const addCache = async (key: string, response: Response) => {
+	if (await isCacheFull()) {
```
* Would the cache ever be full? E.g. say we have 150KB left and we're trying to write a 200KB file. The operation would fail even though the cache isn't full yet.
* Is this check any different from trying to write and throwing when that fails?

bgrgicak replied:
You are correct, this doesn't make any sense. I removed the full cache check in [5e4d336](https://github.com/WordPress/wordpress-playground/pull/1535/commits/5e4d3361f6cfdb1a8b3df185e473d0ab18f961fb).

adamziel commented on packages/php-wasm/web-service-worker/src/worker-caching.ts:
```diff
@@ -0,0 +1,86 @@
+	return await cache.then((c) => c.match(key));
+};
+
+const validHostnames = ['playground.wordpress.net', '127.0.0.1', 'localhost'];
```
Nit: `validHostname` suggests a string that a URL parser would recognize as a hostname. What do you think about `cacheEnabledHosts`?

adamziel replied:
Also, could this be inlined into the `isValidHostname` function below?

bgrgicak replied:
Moved it inline

bgrgicak commented:
https://github.com/WordPress/wordpress-playground/pull/1541 adds a cache version to the Web service worker.
Once this is merged I can update this PR with cache busting and cleaning.

bgrgicak commented:
Offline support is close to done, but I wasn't able to wrap it up this week.

The only remaining item is ensuring that all files required by Playground are cached after the first load. This should take me a day to figure out.
I have the code working already but still need some time to ensure we are loading only the files that we need.

I neglected some of my other work this week so I will take a break from offline support on Monday to catch up and get back to working on this after that.

bgrgicak commented:
> The only remaining item is ensuring that all files required by Playground are cached after the first load. This should take me a day to figure out.

This part is working now.

I discovered another issue. Cache keys are set by pathnames to ensure we precache all assets and can fetch them if the query string changes.
But this is an issue because query strings are lost when a cached response is returned.

adamziel commented:
removing this from the project board in favor of https://github.com/WordPress/wordpress-playground/issues/133 – I also linked the PR to the issue to make the link show up in the card:

<img width="330" src="https://github.com/WordPress/wordpress-playground/assets/205419/d56fd325-3ead-4924-9f35-e7a1415806a4">

bgrgicak commented on packages/playground/remote/src/lib/worker-utils.ts:
```diff
@@ -33,18 +33,23 @@ export type ParsedStartupOptions = {
-	receivedParams.phpExtensions = params.getAll('php-extension');
-	receivedParams.siteSlug = params.get('site-slug') || undefined;
-}
+self.addEventListener('message', (event) => {
+	if (event.data?.type === 'startup-options') {
+		const { startupOptions } = event.data;
+		receivedParams.wpVersion = startupOptions.wpVersion;
+		receivedParams.phpVersion = startupOptions.phpVersion;
+		receivedParams.sapiName = startupOptions.sapiName;
+		receivedParams.storage = startupOptions.storage;
+		receivedParams.phpExtensions = startupOptions.phpExtensions;
+		receivedParams.siteSlug = startupOptions.siteSlug;
+	}
+});
 
+// TODO find a better way to pass the startup options
+while (!receivedParams.wpVersion) {
+	// wait for the next event loop
+	await new Promise((resolve) => setTimeout(resolve, 0));
+}
```
When the worker is loaded from the cache, query strings are in the URL but `self.location` doesn't contain the query string.

I tried modifying the response object, but it's read-only. 
Hash data works in the same way as query strings, so we can't switch to that.

The only solution I found is to pass `startupOptions` as a message to the worker.

The current implementation is a dirty draft solution, but it works. 
I'm now searching for a cleaner solution.

bgrgicak commented on packages/php-wasm/web-service-worker/src/worker-caching.ts:
```diff
@@ -0,0 +1,79 @@
+
+		const cache = await caches.open(this.cacheName);
+
+		// Get the cache manifest and add all the files to the cache
+		const manifestResponse = await fetch('/cache-files.json');
+		const websiteUrls = await manifestResponse.json();
+		await cache.addAll([...websiteUrls, ...['/']]);
```
@adamziel this will download all PHP and WP versions into cache.

Is this ok for v1 or should we limit it to the current WP and PHP version?
If yes, users won't be able to switch versions while offline.

adamziel replied:
I would limit this for v1 to avoid adding 200-300mb of extra transfer. No switching is fine for now, we'll find the right way of downloading the missing assets and communicating that to the user as a part od the webapp redesign issue.

bgrgicak commented on packages/php-wasm/web-service-worker/src/worker-caching.ts:
```diff
@@ -0,0 +1,92 @@
+		}
+
+		/**
+		 * Don't cache PHP files because they are dynamic.
```
We assume these requests go to the server

bgrgicak commented on packages/php-wasm/web-service-worker/src/worker-caching.ts:
```diff
@@ -0,0 +1,92 @@
+		if (url.pathname.endsWith('.php')) {
+			return false;
+		}
+		return ['playground.wordpress.net', '127.0.0.1', 'localhost'].includes(
```
Check if there is a playground.wordpress.net domain

bgrgicak commented on README.md:
````diff
@@ -131,6 +131,26 @@ PHP=7.4 npx @php-wasm/cli -v
+To test the offline support you need to build the website and run a local server:
+
+```bash
+npx nx run playground-website:build:wasm-wordpress-net:development
````
```suggestion
PLAYGROUND_URL=http://localhost:9999 npx nx run playground-website:build:wasm-wordpress-net```

bgrgicak commented on README.md:
````diff
@@ -131,6 +131,26 @@ PHP=7.4 npx @php-wasm/cli -v
+To test the offline support you need to build the website and run a local server:
+
+```bash
+PLAYGROUND_URL=http://localhost:9999 npx nx run playground-website:build:wasm-wordpress-net```
````
```suggestion
PLAYGROUND_URL=http://localhost:9999 npx nx run playground-website:build:wasm-wordpress-net
```

bgrgicak commented on packages/playground/remote/src/lib/worker-caching.ts:
```diff
@@ -0,0 +1,92 @@
+		if (url.pathname.endsWith('.php')) {
+			return false;
+		}
+		return ['playground.wordpress.net', '127.0.0.1', 'localhost'].includes(
```
@adamziel you mentioned that there is a way to not hardcode `'playground.wordpress.net'` but I couldn't find it.

I would like to keep this limited to Playground.WordPress.net for now to be safe, so a hardcoded value seems like the right approach to me.

brandonpayton replied:
Perhaps the target hostname should be an argument to the WorkerCache constructor. Workers have access to `self.location.hostname`, and that value could be passed to the constructor.

bgrgicak replied:
Good point @brandonpayton. I made the change in [b888b97](https://github.com/WordPress/wordpress-playground/pull/1535/commits/b888b97cfbea3e0acf0c69f8192e15c75b22fcb6).

bgrgicak commented on packages/playground/remote/service-worker.ts:
```diff
@@ -27,8 +28,66 @@ if (!(self as any).document) {
 
+const cache = new WorkerCache(buildVersion);
+
+/**
+ * For offline mode to work we need to cache all required assets.
+ *
+ * These assets are listed in the `/assets-required-for-offline-mode.json` file
+ * and contain JavaScript, CSS, and other assets required to load the site without
+ * making any network requests.
+ */
+cache.cacheOfflineModeAssets();
+
+/**
+ * Cleanup old cache.
+ *
+ * We cache data based on `buildVersion` which is updated whenever Playground is built.
+ * So when a new version of Playground is deployed, the service worker will remove the old cache and cache the new assets.
+ *
+ * If your build version doesn't change while developing locally check `buildVersionPlugin` for more details on how it's generated.
+ */
+cache.cleanup();
```
@adamziel I'm [exploring triggering caching on boot, but it's tricky](https://github.com/WordPress/wordpress-playground/pull/1623). Let's trigger caching in the worker for v1 and see how we can improve it later.

brandonpayton replied:
Maybe caching application assets for offline mode is best in the Service Worker. It is part of what Service Workers do.

Then the actual app logic and setup, like the boot process, can build on top of that foundation. The application code can run whether it is taken from local cache or retrieved via HTTP. Putting the source caching logic in boot actually seems strange to me.

bgrgicak replied:
> Maybe caching application assets for offline mode is best in the Service Worker. It is part of what Service Workers do.

I agree. We wanted to keep it in the SW, but wanted to trigger `cacheOfflineModeAssets` from boot.

bgrgicak replied:
I agree, this is service worker specifiv and I don't think this is a wrong place to store the code. 

But triggering `cacheOfflineModeAssets` from boot has some benefits. It would allow boot to control what to cache.
For example, we only cache the current PHP and WP versions, if boot had control over it, we could add support for caching more versions.

adamziel replied:
@bgrgicak I'd still keep it in the service worker. The service worker wouldn't do anything eagerly, it would only look for requests to `php.wasm` and `php.js` and then it would cache the responses. With that in place, the boot flow would only have to trigger a `fetch()`. In fact, it wouldn't be boot specific, we could trigger a few fetch() requests in the UI when the user clicks a "fetch all Playground assets" button.

bgrgicak replied:
> With that in place, the boot flow would only have to trigger a fetch(). 

I will explore this in https://github.com/WordPress/wordpress-playground/pull/1623

bgrgicak replied:
I have it working in https://github.com/WordPress/wordpress-playground/pull/1623.

bgrgicak commented on packages/playground/remote/service-worker.ts:
```diff
@@ -27,8 +28,66 @@ if (!(self as any).document) {
+/**
+ * Handle fetch events and respond with cached assets if available.
+ */
+self.addEventListener('fetch', (event) => {
+	const url = new URL(event.request.url);
+
+	/**
+	 * Don't cache requests to the service worker script itself.
+	 */
+	if (url.pathname.startsWith(self.location.pathname)) {
+		return;
+	}
+
+	/**
+	 * Don't cache requests to scoped URLs or if the referrer URL is scoped.
+	 *
+	 * These requests are made to the PHP Worker Thread and are not static assets.
+	 */
+	if (isURLScoped(url)) {
+		return;
+	}
+	let referrerUrl;
+	try {
+		referrerUrl = new URL(event.request.referrer);
+	} catch (e) {
+		// ignore
+	}
+	if (referrerUrl && isURLScoped(referrerUrl)) {
+		return;
+	}
+
+	/**
+	 * Respond with cached assets if available.
+	 * If the asset is not cached, fetch it from the network and cache it.
+	 */
+	event.respondWith(cache.cachedFetch(event.request));
+});
```
In the future, we could merge the fetch lister from `initializeServiceWorker` with this one.

I didn't want to do it now because I'm not sure if removing `initializeServiceWorker` will break backwards compatibility.

brandonpayton commented on packages/playground/remote/src/lib/worker-caching.ts:
```diff
@@ -0,0 +1,92 @@
+		return await cache.then((c) => c.match(key, { ignoreSearch: true }));
+	};
+
+	isValidHostname = (url: URL) => {
```
There are a lot of valid hostnames for which this will return false. It might be clearer to name the method `shouldCacheForHostname()` or something similar.

bgrgicak replied:
I renamed it to `shouldCacheUrl`.

adamziel commented on packages/php-wasm/web/src/lib/register-service-worker.ts:
```diff
@@ -65,8 +66,14 @@ export async function registerServiceWorker(scope: string, scriptUrl: string) {
+	try {
+		await registration.update();
+	} catch (e) {
+		logger.warn(
```
We know if the user is online, let's rethrow the error if they are.

bgrgicak replied:
Makes sense, done [8003609](https://github.com/WordPress/wordpress-playground/pull/1535/commits/800360900967c861707d10068aae9882b4922a44)

adamziel commented on packages/playground/remote/src/lib/worker-caching.ts:
```diff
@@ -0,0 +1,96 @@
+import { isURLScoped } from '@php-wasm/scopes';
+
+export class WorkerCache {
```
Nit: let's name this after why it's useful, not where it's used e.g. PlaygroundOfflineModeCache

bgrgicak replied:
I renamed it to `OfflineModeCache`. [8003609](https://github.com/WordPress/wordpress-playground/pull/1535/commits/800360900967c861707d10068aae9882b4922a44)

adamziel commented:
On the first sight this looks really good. I'll give this another read during the day – it looks like we're almost there! Thank you @bgrgicak ! ❤️

adamziel commented:
* I loaded Playground
* I checked the "Offline" checkbox in the devtools
* I opened the options modal and selected the "browser storage"
* After the page refreshed, CSS assets wouldn't load
* I refreshed the Playground view (not the entire page), and then the CSS assets loaded

I suspect the issue is because:

* On the the initial load, we expect to load the CSS assets until we backfill the ZIP file
* The server isn't there when we're offline so all these HTTP requests fail
* [window.navigator.onLine here](https://github.com/WordPress/wordpress-playground/blob/60c8c602d5ded6e0b1103a2a4188d495b9d7b069/packages/playground/remote/src/lib/boot-playground-remote.ts#L257) is true because we actually are online.

This seems fine. I don't think we need to fix anything here, let's just document it somewhere where we'll look for the root cause in the future. Perhaps in the offline testing instructions in the README and [here](https://github.com/WordPress/wordpress-playground/blob/60c8c602d5ded6e0b1103a2a4188d495b9d7b069/packages/playground/remote/src/lib/boot-playground-remote.ts#L257)?

adamziel commented on packages/playground/remote/service-worker.ts:
```diff
@@ -26,9 +27,66 @@ if (!(self as any).document) {
+ *
+ * These assets are listed in the `/assets-required-for-offline-mode.json` file
+ * and contain JavaScript, CSS, and other assets required to load the site without
+ * making any network requests.
```
Let's make it clear we skipped the `await` on purpose.

adamziel commented on packages/playground/remote/service-worker.ts:
```diff
@@ -26,9 +27,66 @@ if (!(self as any).document) {
+ *
+ * If your build version doesn't change while developing locally check `buildVersionPlugin` for more details on how it's generated.
+ */
+cache.cleanup();
```
The `.cleanup()` name make it seem like we're evicting the assets we just requested, how about calling it `.removeOutdatedFiles()` and calling if before `.cacheOfflineModeAssets()`?

adamziel replied:
Also, let's make it clear we've skipped the `await` on purpose

adamziel commented on packages/playground/remote/src/lib/offline-mode-cache.ts:
```diff
@@ -0,0 +1,97 @@
+		return await cache.then((c) => c.match(key, { ignoreSearch: true }));
+	};
+
+	shouldCacheUrl = (url: URL) => {
```
This is one eager function. It will cache everything we request on the playground.wordpress.net domain, including `/demos/index.html` and `/client/index.js`. I'm worry about the unintended consequences. Would there be any downside to limiting this to the offline mode assets, PHP builds, and WordPress builds?

bgrgicak replied:
Why not cache these files? They are small and will load faster after refresh if they are cached.
We would also need to maintain a list similar to the one in `listAssetsRequiredForOfflineMode`.

adamziel commented:
This works really well with the "Offline" checkbox in devtools, but it breaks down as soon as I stop the local PHP server:

```
TypeError: Failed to update a ServiceWorker for scope ('http://localhost:9999/') with script ('http://localhost:9999/sw.js'): An unknown error occurred when fetching the script.
```

I think that's because of the feedback I provided earlier about rethrowing the error. Hm. I wonder what would be a useful thing to do there – we do want to throw if we have the internet and just can't fetch the service worker, but we do not want to throw if we're falling back the offline mode when playground.wordpress.net is down. That is a tough one.

Perhaps it's fine to not re-throw the error when we already have a previous version of the service worker in place. What do you think @brandonpayton @bgrgicak ?

adamziel commented:
This works really well in my testing @bgrgicak, yay! 🎉 I was able to create a bunch of new Playgrounds, use the browser storage, preserve my site across multiple page loads and keep updating it as I went. Also, it loads so fast now that everything is cached offline. ❤️ 

I left a few comments, but I don't see any major blockers here. Feel free to merge and address them in a follow-up PR.

adamziel commented:
Actually the missing CSS bothered me and I explored a refactor in https://github.com/WordPress/wordpress-playground/pull/1643 – `navigator.onLine` doesn't seem to be useful, what we really want to know is whether we have the assets in cache.

adamziel commented:
I found one more issue and proposed a fix in https://github.com/WordPress/wordpress-playground/pull/1643

bgrgicak commented:
> I think that's because of the feedback I provided earlier about rethrowing the error. Hm. I wonder what would be a useful thing to do there – we do want to throw if we have the internet and just can't fetch the service worker, but we do not want to throw if we're falling back the offline mode when playground.wordpress.net is down. That is a tough one.

That's interesting. I didn't have many issues with it while testing, but I experienced that `navigator.isOnline` would start ignoring the checkbox after a few reloads. Simply rechecking it worked. 
Let's leave the code as is and just log the error. If we notice issues in "real" offline after merging it we can get back to it.

> Actually the missing CSS bothered me and I explored a refactor in https://github.com/WordPress/wordpress-playground/pull/1643 – navigator.onLine doesn't seem to be useful, what we really want to know is whether we have the assets in cache.

Similarly, this just worked for me, but not relaying on the browser with `isOnline` seems like a more robust approach.
Thank you!

bgrgicak commented:
> This seems fine. I don't think we need to fix anything here, let's just document it somewhere where we'll look for the root cause in the future. Perhaps in the offline testing instructions in the README and [here](https://github.com/WordPress/wordpress-playground/blob/60c8c602d5ded6e0b1103a2a4188d495b9d7b069/packages/playground/remote/src/lib/boot-playground-remote.ts#L257)?

With https://github.com/WordPress/wordpress-playground/pull/1643 we don't need the comment anymore, but I still added a comment for testing offline.

bgrgicak commented:
I addressed the feedback and merged https://github.com/WordPress/wordpress-playground/pull/1643. We should be ready to merge now. Thanks for your help @adamziel @brandonpayton!
